### PR TITLE
Don't use fetch in wasm streaming if it's not available

### PR DESCRIPTION
**Pull request by kripken**

### CODE REVIEW

**Comment by dschuff** [comment]
Is there any browser that implements `WebAssembly.InstantiateStreaming` but not `fetch`?

**Comment by kripken** [comment]
I saw this when debugging in the shell. Perhaps a shell environment shouldn't advertise streaming?

**Comment by dschuff** [comment]
Yeah, `instantiateStreaming` is part of the web bindings rather than the JS bindings. It takes an HTML response object as its argument; and in fact the definition of what the response object even is, comes from the fetch API spec (https://developer.mozilla.org/en-US/docs/Web/API/Response) so I don't think it t even really makes sense for the `instantiateStreaming` API to be there without fetch (or some other shell-specific way to get a response?).

**Comment by kripken** [comment]
Makes sense. I can't seem to find the shell where I saw this, so maybe I was doing something else wrong.

**Comment by kripken** [comment]
I found where this happens, it's on new-enough SpiderMonkey.

@lukewagner, the issue here is that the SM shell has `WebAssembly.instantiateStreaming` defined, but not `fetch`. I was going to work around that in emscripten here, but @dschuff suggested that it didn't make sense for a shell env to provide `instantiateStreaming` (for the same reasons it doesn't provide `fetch`, it's a web thing), so perhaps the proper fix is in the shell. Thoughts?

**Comment by lukewagner** [comment]
So `instantiateStreaming` in SM delegates "can you stream this object?" to the embedding and, in the shell, to allow easy unit testing, I made it accept any `ArrayBuffer` or view (it makes a copy and creates a thread that serves the bytes in a stream at a configurable rate and chunk size).  So you can actually test `instantiateStreaming` in the SM shell if you want to.

**Comment by kripken** [comment]
Thanks @lukewagner. Reopening this PR then, looks like we need this change.

File: src/preamble.js [hunk 1]
```diff
@@ -2314,7 +2314,7 @@ function integrateWasmJS() {
       });
     }
     // Prefer streaming instantiation if available.
-    if (!Module['wasmBinary'] && typeof WebAssembly.instantiateStreaming === 'function' && wasmBinaryFile.indexOf('data:') !== 0) {
+    if (!Module['wasmBinary'] && typeof WebAssembly.instantiateStreaming === 'function' && wasmBinaryFile.indexOf('data:') !== 0 && typeof fetch === 'function') {
```
Comment (dschuff) [inline src/preamble.js]: OK; I'm not necessarily opposed to this, I was just making sure it's needed.
Can we split these conditions onto separate lines now though?

Comment (kripken) [reply src/preamble.js]: Done.

**Pull request merged**